AK/StringFloatingPointConversions.cpp

```diff
@@ -0,0 +1,1126 @@
+    bool is_even = ~mantissa & 1;
+    bool accept_smaller = is_even && all_u_zero;
+    bool accept_larger = is_even || !all_w_zero;
+
+    if (!accept_larger)
+        --w;
+
+    bool all_a_zero = accept_smaller;
+    bool all_b_zero = all_v_zero;
+    int last_digit = 0;
+
+    int exponent10 = skipped_iters - max(-synthetic_exponent, 0);
+
+    while (u / 10 < w / 10) {
+        all_a_zero &= u % 10 == 0;
+        all_b_zero &= last_digit == 0;
+        last_digit = v % 10;
+
+        u /= 10;
+        v /= 10;
+        w /= 10;
+        ++exponent10;
+    }
+    if (all_a_zero) {
+        while (u % 10 == 0) {
+            all_b_zero &= last_digit == 0;
+            last_digit = v % 10;
+
+            u /= 10;
+            v /= 10;
+            w /= 10;
+            ++exponent10;
+        }
+    }
+
+    bool is_tie = all_b_zero && last_digit == 5;
+    bool want_round_down = last_digit < 5 || (is_tie && v % 2 == 0);
+    bool round_down = (want_round_down && (u != v || all_a_zero)) || (v + 1 > w);
+    return { sign, round_down ? v : v + 1, exponent10 };
+}
+
+static u128 multiply(u64 a, u64 b)
+{
+#ifdef __SIZEOF_INT128__
+    unsigned __int128 result = (unsigned __int128)a * b;
+    u64 low = result;
+    u64 high = result >> 64;
+    return u128 { low, high };
+#else
+    return u128 { a }.wide_multiply(u128 { b }).low;
+#endif
+}
+
+template<>
+FloatingPointExponentialForm convert_floating_point_to_decimal_exponential_form<float>(float value)
+{
+    auto multiply_and_shift = [&](u64 operand, u64 multiplier, i32 shift) {
+#ifdef __SIZEOF_INT128__
+        auto result = (unsigned __int128)operand * multiplier;
+#else
+        auto result = multiply(operand, multiplier);
+#endif
+        if (shift < 0)
+            return static_cast<u64>(result << static_cast<u32>(-shift));
+        else
+            return static_cast<u64>(result >> static_cast<u32>(shift));
+    };
+
+    return inner_convert_floating_point_to_decimal_exponential_form(value, multiply_and_shift);
+}
+
+template<>
+FloatingPointExponentialForm convert_floating_point_to_decimal_exponential_form<double>(double value)
+{
+    auto multiply_and_shift = [&](u64 operand, u64 const multiplier[2], i32 shift) {
+        u128 a = multiply(operand, multiplier[0]);
+        u128 b = multiply(operand, multiplier[1]) + a.high();
+        u64 c = a.low();
+
+        if (0 <= shift && shift < 64) {
+            return (c >> shift) | (b << static_cast<u32>(64 - shift)).low();
+        } else if (shift < 0) {
+            return c << static_cast<u32>(-shift);
+        } else {
+            VERIFY(64 <= shift && shift <= 128);
+            return (b >> static_cast<u32>(shift - 64)).low();
+        }
+    };
+
+    return inner_convert_floating_point_to_decimal_exponential_form(value, multiply_and_shift);
+}
+
+// Step 0. Precompute lookup tables for the given floating point type.
+// Lookup tables was generated using the following Python script.
+/*
```

Comment:
I'm not sure if we want to keep this entire code in here as a comment.
But I can't think of a better way to do this so unless someone has a nice idea we can just leave it.